compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -1383,6 +1384,41 @@ impl<'ast, 'ra, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
         }
     }
 
+    fn suggest_range_struct_destructuring(
+        &self,
+        err: &mut Diag<'_>,
+        path: &[Segment],
+        source: PathSource<'_, '_, '_>,
+    ) {
+        // We accept Expr here because range bounds (start..end) are parsed as expressions
+        if !matches!(source, PathSource::Pat | PathSource::TupleStruct(..) | PathSource::Expr(..)) {
+            return;
+        }
+
+        if let Some(pat) = self.diag_metadata.current_pat
+            && let ast::PatKind::Range(Some(start_expr), Some(end_expr), _) = &pat.kind
+            && let (ast::ExprKind::Path(None, start_path), ast::ExprKind::Path(None, end_path)) =
+                (&start_expr.kind, &end_expr.kind)
+            && path.len() == 1
+        {
+            let ident = path[0].ident;
+
+            if (start_path.segments.len() == 1 && start_path.segments[0].ident == ident)
+                || (end_path.segments.len() == 1 && end_path.segments[0].ident == ident)
+            {
+                let start_name = start_path.segments[0].ident;
+                let end_name = end_path.segments[0].ident;
+
+                err.span_suggestion_verbose(
+                    pat.span,
+                    "if you meant to destructure a `Range`, use a struct pattern",
```

Comment:
```suggestion
                    "if you meant to destructure a range, use a struct pattern",
```

or

```suggestion
                    "if you meant to destructure a `{name}`, use a struct pattern",
```

where `name` is `tcx.item_name(range_struct)` (see other comments about accounting for `RangeInclusive` & open ranges). The query `item_name` might not work in `rustc_resolve` (it might hang or crash since the data isn't available) but then there should be alternative `Resolver` APIs for obtaining that.

Reply:
used range